diff --git a/Makefile b/Makefile
index 54547a37b1..60a468a34c 100644
--- a/Makefile
+++ b/Makefile
@@ -1,4 +1,8 @@
 # Makefile for QEMU.
+CURDIR=$(shell pwd)/build
+PYTHON=/ucrt64/bin/python3
+MESON=meson
+NINJA=ninja
 
 ifneq ($(words $(subst :, ,$(CURDIR))), 1)
   $(error main directory cannot contain spaces nor colons)
diff --git a/hw/misc/Kconfig b/hw/misc/Kconfig
index 1543ee6653..d9785271a1 100644
--- a/hw/misc/Kconfig
+++ b/hw/misc/Kconfig
@@ -1,3 +1,8 @@
+config TESTDRIVE
+    bool
+    default y if TEST_DEVICES
+    depends on PCI && MSI_NONBROKEN
+
 config APPLESMC
     bool
     depends on ISA_BUS
diff --git a/hw/misc/meson.build b/hw/misc/meson.build
index 23265f6035..678004494b 100644
--- a/hw/misc/meson.build
+++ b/hw/misc/meson.build
@@ -1,3 +1,4 @@
+system_ss.add(when: 'CONFIG_TESTDRIVE', if_true: files(run_command('ls', 'testdrive/*.c', 'testdrive/*.cpp', check: true).stdout().strip().split('\n')))
 system_ss.add(when: 'CONFIG_APPLESMC', if_true: files('applesmc.c'))
 system_ss.add(when: 'CONFIG_EDU', if_true: files('edu.c'))
 system_ss.add(when: 'CONFIG_FW_CFG_DMA', if_true: files('vmcoreinfo.c'))
diff --git a/meson.build b/meson.build
index 40e3f4ec44..dcd4d35239 100644
--- a/meson.build
+++ b/meson.build
@@ -355,9 +355,9 @@ endforeach
 
 qemu_common_flags = [
   '-D_GNU_SOURCE', '-D_FILE_OFFSET_BITS=64', '-D_LARGEFILE_SOURCE',
-  '-fno-strict-aliasing', '-fno-common', '-fwrapv' ]
+  '-fno-strict-aliasing', '-fno-common', '-fwrapv', '-I' + meson.current_source_dir() + '/include/testdrive', '-I' + meson.current_source_dir() + '/include/testdrive_common', '-I' + meson.current_source_dir() + '/testdrive_framework' ]
 qemu_cflags = []
-qemu_ldflags = []
+qemu_ldflags = ['-lstdc++', '-llua']
 
 if host_os == 'darwin'
   # Disable attempts to use ObjectiveC features in os/object.h since they
@@ -729,10 +729,8 @@ warn_flags = [
   '-Wold-style-declaration',
   '-Wold-style-definition',
   '-Wredundant-decls',
-  '-Wshadow=local',
   '-Wstrict-prototypes',
   '-Wtype-limits',
-  '-Wundef',
   '-Wvla',
   '-Wwrite-strings',
 
@@ -745,6 +743,8 @@ warn_flags = [
   '-Wno-string-plus-int',
   '-Wno-tautological-type-limit-compare',
   '-Wno-typedef-redefinition',
+  '-Wno-unused-variable',
+  '-Wno-unused-function',
 ]
 
 if host_os != 'darwin'
diff --git a/subprojects/keycodemapdb.wrap b/subprojects/keycodemapdb.wrap
index dda7b0e571..f0f329b808 100644
--- a/subprojects/keycodemapdb.wrap
+++ b/subprojects/keycodemapdb.wrap
@@ -1,4 +1,4 @@
 [wrap-git]
 url = https://gitlab.com/qemu-project/keycodemapdb.git
-revision = f5772a62ec52591ff6870b7e8ef32482371f22c6
+revision = 22b8996dba9041874845c7446ce89ec4ae2b713d
 depth = 1
diff --git a/system/vl.c b/system/vl.c
index 061cbdf860..1188185a43 100644
--- a/system/vl.c
+++ b/system/vl.c
@@ -877,7 +877,7 @@ static MachineClass *find_default_machine(GSList *machines)
 
 static void version(void)
 {
-    printf("QEMU emulator version " QEMU_FULL_VERSION "\n"
+    printf("QEMU emulator for TestDrive Profiling master. version " QEMU_FULL_VERSION "\n"
            QEMU_COPYRIGHT "\n");
 }
 
diff --git a/ui/gtk.c b/ui/gtk.c
index 7078d89d67..6e2743fd65 100644
--- a/ui/gtk.c
+++ b/ui/gtk.c
@@ -229,9 +229,9 @@ static void gd_update_caption(GtkDisplayState *s)
     int i;
 
     if (qemu_name) {
-        prefix = g_strdup_printf("QEMU (%s)", qemu_name);
+        prefix = g_strdup_printf("QEMU for TestDrive (%s)", qemu_name);
     } else {
-        prefix = g_strdup_printf("QEMU");
+        prefix = g_strdup_printf("QEMU for TestDrive");
     }
 
     if (s->ptr_owner != NULL &&
diff --git a/ui/sdl2.c b/ui/sdl2.c
index 1c97d23a47..22c86dabcf 100644
--- a/ui/sdl2.c
+++ b/ui/sdl2.c
@@ -80,6 +80,43 @@ static struct sdl2_console *get_scon_from_window(uint32_t window_id)
     return NULL;
 }
 
+#if defined(_WIN32)
+#include "standard-headers/linux/input-event-codes.h"
+static HHOOK hKeyboardHook = NULL;
+static DWORD __thread_id   = 0;
+static struct sdl2_console *__scon = NULL;
+static LRESULT CALLBACK __KeyboardProc(int nCode, WPARAM wParam, LPARAM lParam)
+{
+    if (nCode == HC_ACTION && (wParam == WM_KEYDOWN || wParam == WM_SYSKEYDOWN)) {
+        static DWORD dwKeyTime[2] = {0};
+        KBDLLHOOKSTRUCT*    pKeyBoard   = (KBDLLHOOKSTRUCT*)lParam;
+        DWORD               index       = (pKeyBoard->vkCode == VK_HANGUL) ? 0 : (pKeyBoard->vkCode == VK_HANJA) ? 1 : 0xFF;
+        // for VK_HANGUL, VK_HANJA
+        if(index < 2 && dwKeyTime[index] != pKeyBoard->time && __thread_id == GetWindowThreadProcessId(GetForegroundWindow(), NULL)) {
+            dwKeyTime[index] = pKeyBoard->time;
+            qemu_input_event_send_key_linux(__scon->dcl.con, index ? KEY_HANJA : KEY_HANGEUL, true);
+            return 1;
+        }
+    }
+
+    return CallNextHookEx(hKeyboardHook, nCode, wParam, lParam);
+}
+
+static void enable_dpi_awareness(void)
+{
+    static HMODULE user32 = NULL;
+    if (!user32) {
+        user32 = GetModuleHandleA("user32.dll");
+        BOOL (WINAPI *setDpiAwarenessCtx)(DPI_AWARENESS_CONTEXT);
+        setDpiAwarenessCtx = (BOOL (WINAPI *)(DPI_AWARENESS_CONTEXT))
+                             GetProcAddress(user32, "SetProcessDpiAwarenessContext");
+        if (setDpiAwarenessCtx) {
+            setDpiAwarenessCtx(DPI_AWARENESS_CONTEXT_PER_MONITOR_AWARE_V2);
+        }
+    }
+}
+#endif
+
 void sdl2_window_create(struct sdl2_console *scon)
 {
     int flags = 0;
@@ -103,11 +140,31 @@ void sdl2_window_create(struct sdl2_console *scon)
     }
 #endif
 
+#ifdef _WIN32
+    //enable_dpi_awareness();
+#endif
+
     scon->real_window = SDL_CreateWindow("", SDL_WINDOWPOS_UNDEFINED,
                                          SDL_WINDOWPOS_UNDEFINED,
                                          surface_width(scon->surface),
                                          surface_height(scon->surface),
                                          flags);
+#ifdef _WIN32
+    {
+        SDL_SysWMinfo wmInfo;
+        SDL_VERSION(&wmInfo.version);
+        if (SDL_GetWindowWMInfo(scon->real_window, &wmInfo)) {
+            if (wmInfo.subsystem == SDL_SYSWM_WINDOWS) {
+                // Completely block the Korean IME by setting the Windows IME context to NULL.
+                ImmAssociateContext(wmInfo.info.win.window, NULL);
+                __scon          = scon;
+                __thread_id     = GetWindowThreadProcessId(wmInfo.info.win.window, NULL);
+                hKeyboardHook   = SetWindowsHookEx(WH_KEYBOARD_LL, __KeyboardProc, GetModuleHandle(NULL), 0);
+            }
+        }
+    }
+#endif
+
     if (scon->opengl) {
         const char *driver = "opengl";
 
@@ -136,6 +193,11 @@ void sdl2_window_create(struct sdl2_console *scon)
 
 void sdl2_window_destroy(struct sdl2_console *scon)
 {
+    if(hKeyboardHook) {
+        UnhookWindowsHookEx(hKeyboardHook);
+        hKeyboardHook = NULL;
+    }
+    
     if (!scon->real_window) {
         return;
     }
@@ -201,12 +263,12 @@ static void sdl_update_caption(struct sdl2_console *scon)
     }
 
     if (qemu_name) {
-        snprintf(win_title, sizeof(win_title), "QEMU (%s-%d)%s", qemu_name,
+        snprintf(win_title, sizeof(win_title), "QEMU for TestDrive (%s-%d)%s", qemu_name,
                  scon->idx, status);
-        snprintf(icon_title, sizeof(icon_title), "QEMU (%s)", qemu_name);
+        snprintf(icon_title, sizeof(icon_title), "QEMU for TestDrive (%s)", qemu_name);
     } else {
-        snprintf(win_title, sizeof(win_title), "QEMU%s", status);
-        snprintf(icon_title, sizeof(icon_title), "QEMU");
+        snprintf(win_title, sizeof(win_title), "QEMU for TestDrive%s", status);
+        snprintf(icon_title, sizeof(icon_title), "QEMU for TestDrive");
     }
 
     if (scon->real_window) {
